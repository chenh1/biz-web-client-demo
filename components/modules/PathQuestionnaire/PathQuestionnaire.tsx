"use client";

import type { PathQuestion } from "@/graphql/queries/pathQuestionnaire";
import type { Answer } from "@/graphql/queries/examQuestions";
import React, { useState, useEffect } from "react"
import { Box } from "../../containers/Box/Box";
import { Heading } from "../../core/Heading/Heading";
import { Button } from "../../core/Button/Button";
import { useRouter } from "next/navigation";
import { Loader } from "../Loader/Loader";

interface PathQuestionnaireProps {
  questions: PathQuestion[]
}

export const PathQuestionnaire: React.FC<PathQuestionnaireProps> = ({ questions }: PathQuestionnaireProps) => {
  const router = useRouter();
  const [ onQuestion, setOnQuestion ] = useState<number>(0);
  const [ answers, setAnswers ] = useState<Answer[]>([]);
  const [ confirmStart, setConfirmStart ] = useState<boolean>(false);

  const currentQuestion = questions[onQuestion];
  
  useEffect(() => {
    const qs = answers.map(({ key }, i) => `q${i + 1}=${key}`);
    if (onQuestion === questions.length) {
      router.push(`/dashboard?${qs.join('&')}`);
    }
  }, [onQuestion])

  return (
    <Box padding="0" className="justify-center">
      <Box gap="md" padding="0" className="max-w-xl">
        {!confirmStart ?
          <>
            <Heading type="h4">Before we get started, we'd like to get to know you a little bit. This way, we'll better understand which learning modules suit you!</Heading>
            <Button 
              data-cy={`path-questionnaire-confirm`}
              onClick={() => setConfirmStart(true)}
              primary
              label="Let's go!"
            />
          </>
          :
          <>
            <Heading type="h4">{currentQuestion?.question}</Heading>
            <Box gap="sm" padding="0">
              {currentQuestion?.answers?.map((datum, i) => (
                <Button 
                  key={i}
                  textAlign="left"
                  data-cy={`path-questionnaire-answer-${i}`}
                  onClick={() => {
                    setAnswers([...answers, datum]);
                    setOnQuestion(onQuestion + 1);
                  }}
                  label={datum?.answer}
                />
              ))}
            </Box>
            {onQuestion + 1 <= questions?.length && <Heading type="h4">{`${onQuestion + 1} of ${questions?.length}`}</Heading>}
          </>
        }
        {onQuestion === questions.length &&
          <Loader/>
        }
      </Box>
    </Box>
  )
}