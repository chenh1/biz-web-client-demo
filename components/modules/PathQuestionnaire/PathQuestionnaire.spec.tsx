import { PathQuestionnaire } from "./PathQuestionnaire";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("next/navigation", () => ({
  useRouter: jest.fn()
}));

import { useRouter } from "next/navigation";
const mockedUseRouter = useRouter as jest.Mock;
const push = jest.fn();

mockedUseRouter.mockImplementation(() => ({
  prefetch: () => null,
  push: push
}));

it('renders PathQuestionnaire module', async () => {
  const { getByTestId, asFragment } = render(
    <PathQuestionnaire
      questions={[
        { question: "Are you happy?", answers: [{ key: "yes", answer: "yes" }, { key: "no", answer: "no" }] },
        { question: "Are you content?", answers: [{ key: "yes", answer: "yes" }, { key: "no", answer: "no" }] },
      ]}
    />
  )

  expect(asFragment()).toMatchSnapshot();
});

it('appends answers to query string for parsing of archetypes', async () => {
  const { getAllByTestId, asFragment } = render(
    <PathQuestionnaire
      questions={[
        { question: "Are you happy?", answers: [{ key: "yes", answer: "yes" }, { key: "no", answer: "no" }] },
        { question: "Are you content?", answers: [{ key: "yes", answer: "yes" }, { key: "no", answer: "no" }] },
      ]}
    />
  )

  fireEvent.click(getAllByTestId("button")[0])
  fireEvent.click(getAllByTestId("button")[0])
  fireEvent.click(getAllByTestId("button")[0])
  
  expect(push).toHaveBeenCalledWith("/dashboard?q1=yes&q2=yes")
});
