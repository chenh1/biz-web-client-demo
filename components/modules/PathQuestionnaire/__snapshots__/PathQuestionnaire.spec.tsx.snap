// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PathQuestionnaire module 1`] = `
<DocumentFragment>
  <div
    class=" grid  gap-4  p-2 justify-center"
  >
    <div
      class=" grid  gap-8  p-2 max-w-xl"
    >
      <h4
        class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md "
      >
        Before we get started, we'd like to get to know you a little bit. This way, we'll better understand which learning modules suit you!
      </h4>
      <button
        class="   w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-max-white px-5 py-2.5   bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red"
        data-cy="path-questionnaire-confirm"
        data-testid="button"
        type="button"
      >
        <div
          class="bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red rounded-lg absolute left-0 top-0 w-full h-full "
        />
        <span
          class="pointer-events-none relative z-10"
        >
          Let's go!
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
